core/consensus/pow/primitives/src/lib.rs

```diff
@@ -0,0 +1,29 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use substrate_client::decl_runtime_apis;
+use rstd::vec::Vec;
+use sr_primitives::ConsensusEngineId;
+use primitives::H256;
+
+/// The `ConsensusEngineId` of PoW.
+pub const POW_ENGINE_ID: ConsensusEngineId = [b'p', b'o', b'w', b'_'];
+
+/// Type of difficulty.
+pub type Difficulty = u128;
+
+/// Type of seal.
+pub type Seal = Vec<u8>;
+
+decl_runtime_apis! {
+	/// API necessary for block authorship with Proof of Work.
+	pub trait PowApi {
+		/// Verify a given proof of work against the current difficulty.
+		/// Note that `pre_hash` is always a hash of a direct child.
+		///
+		/// Returns the current difficulty.
+		fn verify(pre_hash: &H256, seal: &Seal) -> Option<Difficulty>;
```

Comment:
I see - instead of using a particular algorithm we are delegating all of it to the runtime. Is it a pure function? This will definitely complicate things for light clients (and parallel verification queues, which are a kind of light client), if the expectation is that it will be called at the parent's runtime. If that's not the expectation, then how do you do algorithm changes?

Reply:
> This will definitely complicate things for light clients (and parallel verification queues, which are a kind of light client), if the expectation is that it will be called at the parent's runtime.

Yes. The expectation is that it will be called at the parent's runtime. This is a "pure function" in the sense that it will not modify anything in the parent runtime. However, usually it needs to take in parent storage information for it's difficulty adjustment algorithm (parent block's difficulty).

Reply:
TBH, I think it isn't usable for anything except the most basic circumstances to take such an approach. It makes light clients and warp sync much more difficult and less space-efficient.

Reply:
TBH I don't think we can create a light client implementation that fits all consensus engines, so I just assumed that we temporarily don't consider light client here.

Reply:
> I don't think we can create one light client implementation that fits all consensus engines

I agree - pluggable consensus is not designed to do that. But you'll have a hard time writing a PoW light client with this design. That's unfortunate, since PoW should be one of the easiest things to write a light client for (as it is not intertwined with the state)

Reply:
> But you'll have a hard time writing a PoW light client with this design. That's unfortunate, since PoW should be one of the easiest things to write a light client for (as consensus state is not intertwined with the state machine)

Consider, if we design a light client for this PoW, can you explain more why you think it would be hard? If we just add a digest item of the hash of the current runtime code, the light client can then make assumptions about the PoW algorithm. This then is effectively the same as defining separate consensus engines for PoW and use markers to change them on the fly.

Reply:
The particular use case I have in mind is that a lot of PoW blockchains like to "tweak" their algorithm once in a while. If we define separate engines for each PoW algorithm, then there'll never be an end that we can support all of them -- it will just be too many. Besides, any tweak in the algorithm will require "hard fork" (in the sense that one must update the Substrate version running). This kind of goes against the goals of Substrate's "online upgrade".

Reply:
Yes, I understand the goal.

> If we just add a digest item of the hash of the current runtime code, the light client can then make assumptions about the PoW algorithm

Right - either this, or having the runtime include some other WASM code for PoW verification (where light clients can download this module when it changes, and the full node could execute it using the sandbox utilities). We also need to write some code for full nodes to ensure that the hash actually matches the runtime state at all times.

Reply:
> The particular use case I have in mind is that a lot of PoW blockchains like to "tweak" their algorithm once in a while

Thought about this a bit more. How do you reason about fork-choice? What happens when a chain switches from longest-chain PoW to GHOST-based or vice-versa? What about GHOST-based to GHOST-based? What about something 

> If we define separate engines for each PoW algorithm, then there'll never be an end that we can support all of them -- it will just be too many. 

It makes some sense to define a skeleton consensus crate that is instantiatable by a `T: PoWAlgorithm`.

Reply:
> Thought about this a bit more. How do you reason about fork-choice? What happens when a chain switches from longest-chain PoW to GHOST-based or vice-versa? What about GHOST-based to GHOST-based? What about something

Then they will *have to* switch to a different consensus engine. Here's assumptions we make for the current engine:

* Definition of "difficulty" and "timestamp".
* Each block has an associated seal and difficulty.
* Chain selection chooses highest total difficulty.
* We make no assumption how runtime defines difficulty or validates its seal.

> It makes some sense to define a skeleton consensus crate that is instantiatable by a T: PoWAlgorithm.

Anyway, I think we can just support both -- light client "friendly" version where algorithms are defined outside of runtime, and light client "unfriendly" version where they are defined within runtime by using this `T: PowAlgorithm` generalization. I think that'll just resolve this once and for all. xD

Reply:
How do you envision light-clients working, where the PoW algorithm can change on any block? Since there isn't any delay in the PoW algorithm changing, all a bad miner needs to do is give light clients 1 block and then they can be tricked into following that chain forever (by giving it a difficulty that grows faster than the legitimate chain).

I don't think it's a good idea to give people the option to deploy chains that will _never be safe for light clients_ without a strong disclaimer. Like naming the crate `full_clients_only_somewhat_dynamic_pow`.

Reply:
Actually, you can extend this to full nodes too. Let's say you could have 2 forks which each pass a different runtime upgrade via a governance process. Runtime upgrade A does a minor tweak of the process, and upgrade B does a big change, that causes difficulty to skyrocket afterwards. Fork B is done in secret by a cartel of malicious miners, who push through a unanimous coin-vote to pass upgrade B. They then mine on top of it a bit more and release fork B to the network. Now fork B will be considered higher total difficulty than fork A.

They could also bundle any other changes in the runtime upgrade that they wanted.

Reply:
(you don't even need fork A to do a runtime upgrade; it's irrelevant. all that matters it that a small minority of miners can change things in a secret fork and then overwhelm the honest fork. that's not the security model of PoW)

Reply:
> Runtime upgrade A does a minor tweak of the process, and upgrade B does a big change, that causes difficulty to skyrocket afterwards. Fork B is done in secret by a cartel of malicious miners, who push through a unanimous coin-vote to pass upgrade B. They then mine on top of it a bit more and release fork B to the network. Now fork B will be considered higher total difficulty than fork A.

I don't think this is any worse than PoS process. To actually do this, the attacker would need to trick the governance process so that its fork B somehow passed. This is rather a vulnerability in onchain goverance, than PoW engine.

Anyway, I wouldn't mind if we don't support `PowRuntimeAlgorithm` and always stick to fixed PoW algorithm for a chain. Maybe we just remove it?

Reply:
> Anyway, I wouldn't mind if we don't support PowRuntimeAlgorithm and always stick to fixed PoW algorithm for a chain. Maybe we just remove it?

I recommend it as the safest option for now

I'm happy to revisit the runtime-based algorithm later on, but it appears to need more research about unification of fork-choice rules.

> I don't think this is any worse than PoS process. To actually do this, the attacker would need to trick the governance process so that its fork B somehow passed. This is rather a vulnerability in onchain goverance, than PoW engine.

That argument can be made, but it is in fact the case with SRML governance (you can pass anything with unanimity in `srml-democracy`, and a secret fork can contain only "for" transactions.

It's not the same situation as BABE/GRANDPA since you have finality that prevents reorganization, although there might be a similar vector on BABE that I'd like to confirm before posting here.

Reply:
Hmm, I’m actually having a feeling that the issue may be worse. Even with finality, if last finalised block is old enough than current chain head, this attack can still be carried out, right? Because a minority fork can arbitrarily change the validator set (and thus fork choice) in a short period of time (if epoch is less than voting period).

Reply:
I removed `PowRuntimeApi` in `745d604b02b19ef98aa89ca488941d12b299bd53` temporarily. I think the actual problem lies deeper than in the runtime PoW engine, and we may want to explore to see whether we can do something about it. From my understanding, it looks like the governance/democracy module may be *unsafe* for any consensus engine that requires runtime data. Two potential ways I see moving forward:

* Keep it strictly that consensus engine cannot access runtime data. This is the same way as Serenity -- change consensus using hard fork, change runtime execution using WebAssembly.
* Find some ways to make governance module safe for those consensus engines.

Reply:
I created a new issue discussing swappable PoW (https://github.com/paritytech/substrate/issues/3517). As we mostly reached conclusion that we're just not touching that subject on this PR, going to mark this as resolved.